test/test_elbo.jl

```diff
@@ -314,6 +318,35 @@ function test_num_allowed_sd()
 end
 
 
+function test_elbo_supports_dual_numbers()
+    _, ea, _ = gen_two_body_dataset()
+
+    # create elbo arguments of type Float64 that compute the gradient but not the hessian
+    ea0 = ElboArgs(ea.images, ea.vp, ea.patches, ea.active_sources, calculate_hessian=false)
+
+    # create elbo arguments of the Dual number type, that compute the gradient but not
+    # the hessian (it doesn't matter that the "perturbation" are all zero, it's just
+    # for testing the speed and verifying that it works)
+    P = length(ea0.vp[1])
+
+    # `1` "perterbation" per dual number is enough for a hessian-vector mulitiply,
+    T = ForwardDiff.Dual{1, Float64}
+    vp = Vector{T}[zeros(T, P) for s=1:ea0.S]
+    for s=1:ea0.S
+        vp[s][:] = ea0.vp[s][:]
+    end
+    ea1 = ElboArgs(ea0.images, vp, ea0.patches, ea0.active_sources, calculate_hessian=false)
+
+    # evaluate the elbo for both argument types
+    DeterministicVI.elbo(ea0)
+
+    chunksize = ForwardDiff.pickchunksize(length(Model.ids))
+    kl_helper = KLHelper(Dual{chunksize, Dual{1, Float64}})
```

Comment:
You shouldn't allocate `KLHelper`s inside of other functions, only in top-level scope. For example, I'm pretty sure this can never pass on v0.6 since the world-ages of the functions created by the KLHelper are too young compared to the world-age of the calling scope.